biblib/views/document_view.py

```diff
@@ -44,20 +45,79 @@ def add_document_to_library(cls, library_id, document_data):
+                current_app.logger.error("Failed to retrieve bibcodes with error: {}".format(solr_resp.get("error")))
+                output_dict = {"error": solr_resp.get("error"), "number_added": 0, "status": status_code}
+                valid_bibcodes = []
+                return output_dict
+            else:
+                #If SOLR query succeeds generate list of valid bibcodes from response
+                valid_bibcodes = [doc.get('bibcode') for doc in solr_resp.get('docs', {})]
+                current_app.logger.debug("Found the following valid bibcodes: {}".format(valid_bibcodes))
+            
+            if valid_bibcodes:
+                #Add all valid bibcodes to library
+                library.add_bibcodes(valid_bibcodes)
+
+                session.add(library)
+                session.commit()
+
+                current_app.logger.info('Added: {0} to {1}'.format(
+                    valid_bibcodes,
+                    library_id)
+                )
+                
+                current_app.logger.debug('Added: {0} is now {1}'.format(
+                    valid_bibcodes,
+                    library.bibcode)
+                )
+            
             end_length = len(library.bibcode)
 
-            return end_length - start_length
+            #Generate a list of invalid bibcodes
+            invalid_bibcodes = list(set(document_data['bibcode']) - set(valid_bibcodes))
+            
+            #Generate output that contains the number added and the number of invalid bibcodes.
+            output_dict = {"number_added": end_length - start_length}
+            if invalid_bibcodes: output_dict['invalid_bibcodes'] = invalid_bibcodes
+            
+            return  output_dict
+
 
+    @classmethod
+    def query_valid_bibcodes(cls, input_bibcodes):
+        """
+        Takes a list of input bibcodes and validates there existence in ADS
+        through the API. Calls either standard search or bigquery depending 
+        on the query length.
+        """
+        bigquery_min = current_app.config.get('BIBLIB_SOLR_BIG_QUERY_MIN', 10)
+        if len(input_bibcodes) < bigquery_min:
+            try:
+                response = cls.standard_ADS_bibcode_query(input_bibcodes)
+                solr_resp = response.json()
+                status = response.status_code
+            except Exception as err:
+                current_app.logger.error("Failed to collect valid bibcodes from input due to internal error: {}.".format(err))
+                solr_resp = {"response": {"error": "An internal error occurred when querying SOLR. Please try again later."}}
+                status = 500
+        else:
+            try:
+                #For calls to bigquery, we limit the number of rows allowed in config.
+                response = cls.solr_big_query(input_bibcodes, rows=min(len(input_bibcodes), current_app.config.get('BIGQUERY_MAX_ROWS', len(input_bibcodes))))
```

Comment:
Somewhere, we should check that `input_bibcodes` is less than or equal to `BIGQUERY_MAX_ROWS` (either here or in `solr_big_query` or in the calling function...) and do something (paginate?) if so

Reply:
I made some initial changes so that add_document_to_library() requests based on the max number of rows allowed and then pages through the query. I currently have it set up to define the number of pages based on `input_bibcodes`, but I want to look at that a little more because it could cause some issues if a large number of the input bibcodes turn out to be invalid.

Reply:
I added a check on the query to make sure every request actually returns bibcodes. If it doesn't it breaks out of the loop. That should keep biblib from making a bunch of extra queries in the event there are more than a single page worth of invalid bibcodes.

Reply:
Okay, I have a test that checks for multiple calls to Bigquery and iterates through the bibcodes to emulate paging through bigquery responses.